src/commands/classified/external.rs

```diff
@@ -50,158 +49,243 @@ impl Decoder for LinesCodec {
+    trace!(target: "nu::run::external", "cwd = {:?}", context.shell_manager.path());
 
-    if let Some(stdin) = stdin {
-        process = process.stdin(stdin);
+    if !is_last {
+        process = process.stdout(subprocess::Redirection::Pipe);
+        trace!(target: "nu::run::external", "set up stdout pipe");
     }
 
+    // NOTE can't use a subprocess::make_pipe here. There seems to be no way to close stdin
+    //      after doing so, which makes certain commands block forever (e.g., cat).
+    process = process.stdin(subprocess::Redirection::Pipe);
     trace!(target: "nu::run::external", "set up stdin pipe");
+
     trace!(target: "nu::run::external", "built process {:?}", process);
 
     let popen = process.popen();
+    if let Ok(mut popen) = popen {
+        let _ = popen.detach();
+        let mut stdin_write = popen.stdin.take().unwrap();
 
-    trace!(target: "nu::run::external", "next = {:?}", stream_next);
+        // Instead of writing everything into the pipe, which may(?) overflow, we write small
+        // amounts at a time.
+        let name_tag2 = name_tag.clone();
+        let write_to_stdin_stream = async_stream! {
+            while let Some(item) = input.next().await {
+                match item.value {
+                    UntaggedValue::Primitive(Primitive::String(s)) => {
+                        stdin_write.write(s.as_bytes());
+                        stdin_write.write(b"\n");
+                        yield Ok(());
+                    },
+                    UntaggedValue::Primitive(Primitive::Line(line)) => {
+                        stdin_write.write(line.as_bytes());
+                        yield Ok(());
+                    },
 
-    let name_tag = command.name_tag.clone();
-    if let Ok(mut popen) = popen {
-        match stream_next {
-            StreamNext::Last => {
-                popen.detach();
+                    // TODO perhaps serialize some other primitives?
+                    _ => {
+                        yield Err(ShellError::labeled_error(
+                            "Unexpected stdin type found in input stream",
+                            "unexpected type",
+                            name_tag2.clone()
+                        ));
+                        break;
+                    }
+                }
+            }
+        };
+
+        let stream = async_stream! {
+            pin_mut!(write_to_stdin_stream);
+
+            if is_last {
+                let mut still_has_stdin = true;
                 loop {
-                    match popen.poll() {
-                        None => {
-                            std::thread::sleep(std::time::Duration::new(0, 100_000_000));
-                        }
-                        _ => {
-                            let _ = popen.terminate();
-                            break;
+                    if let Some(status) = popen.poll() {
+                        // TODO "kill" write_to_stdin_stream?
```

Comment:
If you can merge into one async_stream, then you could test to see if it was time to break out of the input loop

Reply:
Good point. We'd still technically have `input` to deal with, but I'm guessing I'm worrying more than I need to. Everything will eventually get `Drop`-ed, and the appropriate cleanup will happen.

Reply:
@jonathandturner I just realized a benefit to separating "write to stdin" and "yield from stdout" into separate async steps.

Imagine a process that takes some input, and then goes through a lengthy computation, but emits things to stdout throughout that process. If we have to wait to feed it all of stdin, we delay getting data from its stdout to the next step in the pipeline.

A scenario like that which may not be super contrived:

```
... | curl_each_url | from-json | ...
```

If we have to wait for all of the `curl`s to complete, we "stall" the pipeline.

Reply:
Although, that assumes an unbuffered pipe...

Reply:
Let's assume buffered pipes and ignore the above example. We can cross the bridge of splitting the two processes when we get actual reports of it being a problem 🙂